CCL draft RFC

dhuseby opened the pull request:
This is the draft RFC for the Cryptographic Construction Language (CCL) for encoding cryptographic constructions in a self-describing way that is independent from the underlying crypto library.

Signed-off-by: Dave Huseby <dhuseby@linuxfoundation.org>

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+endosring signature must be set up with reads of the data and reads of the
+previous signature data before calculating the hash to compare.
+
+Any number of digital signatures can be generated and appeneded to the existing
```
s/appeneded/appended

dhuseby replied:
fixed

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+
+This would leave the stack with the JSON bytes before and after the signature
+line on the stack ready to be parsed, canonicalized, and then hashed. If the
+JSON is assumed to already be canoncicalized, then the hash can be directly
```
s/canoncicalized/canoncialized/

dhuseby replied:
fixed

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+
+By storing cryptographic constructs in "functional" form, we abstract away the
+underlying cryptographic privitive/operation providers as well as provide a
+standardized language for describing arbitrarily complex cryptogrphic
```
s/cryptogrphic/cryptographic/

dhuseby replied:
fixed

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+* Should we focus on making sure CCL is non-Turing-complete? These are
+  essentially small "smart contracts" and therefore they are safety critical.
+  Non-Turing-completeness would allow for the creation of static analysis tools
+  and determistic evaluation looking for undesirable corner case possibilities.
```
s/determistic/deterministic/

dhuseby replied:
fixed

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+The first job of mapping abstract CCL to JSON is to decide the string constants
+to use for the supported encoding types, encryption algorithms, hashing
+algorithms, and opcodes. Below are lists of constants in CCL and their string
+constant equivilents in JSON.
```
s/equivilents/equivalents/

dhuseby replied:
fixed

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+stack and evaluates it as a boolean argument.
+
+In the case of `IF-ELSE-THEN` if the argument is true then the script between
+`IF` and `ELSE` is excuted followed by jumpting to the script after the `THEN`.
```
s/excuted/executed/
s/jumpting/jumping/

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+verified and does not require the full parsing and manipulation of the JSON
+document.
+
+For insance, in Secure Scuttlebutt, digital signatures are stored as:
```
s/insance/instance/

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+For insance, in Secure Scuttlebutt, digital signatures are stored as:
+`"signature": "<base64>.sig.ed25519"`. The specification says to parse the
+JSON, remove the signature field, then canoncialize the JSON before hashing.
+With CCL, if the signature data starts at offest 152 and is 100 bytes long,
```
s/offest/offset/

danintel commented on 0000-crypto-construct-language.md:
````diff
@@ -0,0 +1,761 @@
+```
+
+Comment lines being with a single `/` character and come before the command and
+stack documentation to give explanations of the parametes int he stack diagram.
````
s/parametes/parameters/

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+[motivation]: #motivation
+
+By storing cryptographic constructs in "functional" form, we abstract away the
+underlying cryptographic privitive/operation providers as well as provide a
```
s/privitive/primitive/

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+Comment lines being with a single `/` character and come before the command and
+stack documentation to give explanations of the parametes int he stack diagram.
+The stack diagram follow the command---in this case `=`---and consists of a
+right parenthesis `(` followed by the parameters poped off of the stack
```
s/poped/popped/

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+binding and key rotation operations. Those operations typically have one or
+more ways that a controller can prove control over an identity and/or force
+a rotation away from a compromised key. This RFC seeks to standardize the
+language we use to describe those constructs so that regarless of the
```
s/regarless/regardless/

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+So far no good proposals have been made to adapt Git to multi-sig and Secure
+Scuttlebutt has just rejected multi-sig altogether for now. There have been
+some attempts at adapting Linked Data Signatures used in DIDs to support
+multi-sigs. What was proposed is a little similary to CCL but is clunky and
```
s/similary/similarly/ (or similar?)
Or
s/little similary/slightly similar/ ?

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+
+The first token is the Base64 encoded public key data followed by the
+identifier for the encoding scheme (e.g. "BASE64") and lastly an opcode
+executing the decode operation. Sincel CCL is processed using a stack machine
```
s/Sincel/Since/

dhuseby replied:
fixed

danintel commented on 0000-crypto-construct-language.md:
````diff
@@ -0,0 +1,761 @@
+```
+
+This is a parallel multi-sig. The first part is a normal CCL digital signature
+valudidation. After the `=` opcode, the stack will either have a TRUE on top if
````
s/valudidation/validation/

dhuseby replied:
fixed

danintel commented on 0000-crypto-construct-language.md:
````diff
@@ -0,0 +1,761 @@
+```
+
+Let's say another signer wants to add a parallel signature to the one above.
+They would calculate their own signature and append it using `IF-ELSE-THEN` to
````
Maybe this is some pre-existing convention somewhere (Forth? Stack machine?), but the IF-ELSE-THEN construct is confusing. I would think a IF-ELSE-FI (shell script) or IF-ELSE-ENDIF (ccp) would be less confusing.

dhuseby replied:
It is Forth: https://www.forth.com/starting-forth/4-conditional-if-then-statements/

I'm not married to the convention and we can totally adopt something more straightforward.

dhuseby replied:
I changed it to IF-ELSE-FI. Thanks for the suggestion.

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+
+## Encoding Formats
+
+The first version of CCL supports the following encoding types:
```
These are good choices (I'm not sure Base58 is needed, but it reduces confusion). I think you also need to specify the encoding a little bit more. For example, with Base64 I assume it is not with BEGIN and END lines--it is all one string.  But is it enclosed in quotes? Or just delimited with whitespace?

dhuseby replied:
I included more description on what I mean for each encoding type. I also included Base58 to show that this is intended to be extensible and application neutral. I also added language saying that Base58Check is a complex construct that can be stored in CCL as a script that unpacks and checks the hashes and decodes the Base58 pieces leaving the payload string in binary form on the stack.

danintel commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,761 @@
+rules need to be serialized in a form that is machine parsable and machine
+executable with context specific references to files and commits to read data
+used in the construct creation. This RFC again seeks to standardize the
+language we use to describe those rules.
```
You mention these are for describing rules, but in the examples below you use hard-coded values for keys and signatures (implying a specific execution stream that will be interpreted and executed by an interpreter). Which is it? Both?

Maybe you can show such an example rule for MAINTAINERS file changes encoded in CCL.

dhuseby replied:
addressed.

dhuseby commented:
@danintel There are three commits in this PR. I addressed your feedback in the second commit---https://github.com/hyperledger/ursa-rfcs/pull/17/commits/4cf639161276adf9f2de16fac9bcd7a508125b85---and then later added some minor clean ups in the third commit.

danintel commented:
@dhuseby: I saw the multiple commits.  Usually the commits are aggregated under "Files changed".  I'll look again.

jaromil commented:
If I understand well your value proposition, you may want to acknowledge Zencode as prior art:
- https://dev.zenroom.org/Zencode
- https://zenroom.org
We already use it connected to an hyperledger product.

dhuseby commented:
> If I understand well your value proposition, you may want to acknowledge Zencode as prior art:
> 
>     * https://dev.zenroom.org/Zencode
> 
>     * https://zenroom.org
>       We already use it connected to an hyperledger product.

I have been reading LangSec papers for years and, like zenroom, a lot of the thinking is inspired from there. However, after looking at Zenroom and Zencode, it isn't an inspiration for CCL. CCL isn't about the VM, it's more about a uniform way to serialize crypto constructs. Crypto constructs almost always consist of some data and some implied processing of that data to validate, verify, authenticate, encrypt, decrypt, etc. CCL acknowledges that fact and explicitly encodes the processing along with the data. Using a stack machine as the "VM" for executing the CCL encoded constructs just simplifies things and drastically limits attack surface area by allowing for a non-Turing-complete deterministic DSL to be used.

jaromil commented:
Thanks for the clarification! I'll be following the CCL specification with great interest.

mikelodder7 commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,917 @@
+* RSA
+* XSalsa20Poly1305
+* AES
+
```
No `XChaCha20Poly1305`?

mikelodder7 commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,917 @@
+* SHA512-256
+* Blake2b
+* Argon2id
+
```
`Keccak`?

mikelodder7 commented on 0000-crypto-construct-language.md:
```diff
@@ -0,0 +1,917 @@
+* RSA - `RSA`
+* XSalsa20Poly1305 - `XSalsa20Poly1305`
+* AES - `AES`
+
```
Do we want to support any `ECIES` methods like libsodium's secretbox and sealedbox?

dhuseby replied:
XSalsa20Poly1305 is libsodium's SecretBox. In the v0.1.0 version I left out the SealedBox but if you look at the CCLang repo you'll see that it would be trivial to add new construct primitives.

mikelodder7 commented:
Love this idea. Just some minor things to include and I think we’ve got a good start

dhuseby commented:
I just published v0.1.0 of my CCLang implementation. I scaled back on some of the algorithms to get a MVP. The repo is here: https://github.com/dhuseby/cclang and the crate page is here: https://crates.io/crates/cclang

dhuseby commented:
My hope is that CCLang will stop using sodiumoxide as it's crypto library and instead be built into Ursa and integrated with Ursa's build system so that depending on how Ursa is built, the CCLang algorithms and underlying primitives are set up accordingly. This would allow consumers of Ursa to use CCLang in FIPS mode or "open crypto" mode or with any of the restricted sets of algorithms and underlying crypto libraries.

dcmiddle commented:
I gather this should be closed now?
